### PR TITLE
Throw custom exception in NoteTraits if notes are disabled

### LINKED ISSUE #6703: Note creation attempt causes fatal error if `woocommerce_admin_disabled` is true
**Describe the bug**

When trying to create a note using `NoteTraits`, if WC Admin is disabled using this filter:

```php
add_filter( 'woocommerce_admin_disabled', '__return_true' );
```

The note creation will fail with a fatal error.

I ran into this issue while working on https://github.com/Automattic/woocommerce-services/pull/2402.

This has to be worked around by users of the `Automattic\WooCommerce\Admin\Notes\NoteTraits` trait or note classes by checking if either:

* `woocommerce_admin_disabled` is true,
* the `admin-note` data store exists.

It would be nice if users of the trait/classes would not have to worry about whether the data store is defined (which I guess is a bit of WC Admin's implementation detail) when trying to create a note.

Maybe WC Admin could catch `WC_Data_Store::load( 'admin-note' )` exceptions (or check for the data store existence otherwise) in each call to the trait's methods? I see this was already implemented for one note in #6532.

**To Reproduce**

1. Add this to a test site:

    ```php
    add_filter( 'woocommerce_admin_disabled', '__return_true' );
    ```
1. Try to create a note using the trait during the `plugins_loaded` hook.

**Expected behavior**

A fatal error is not thrown regardless of whether the `admin-note` data store is defined. Maybe a `WP_Error` or a custom exception could be thrown instead?

### PR DESCRIPTION
**Pull request by waclawjacek**
Fixes #6703.

When WC Admin is disabled, notes cannot be created using `NoteTraits` due to the `admin-note` data store not being registered.

This PR introduces a custom exception that is thrown in `NoteTraits` whenever WC Admin is disabled and any of the trait's methods that depend on the data store is called (as of this moment, these are all of the methods except for `wc_admin_active_for()`).

I did not add any test instructions to `/TESTING-INSTRUCTIONS.md` because it seems like this is not a kind of test that should be added there. If this is incorrect, please let me know!

### Detailed test instructions:

- Try to use any of the trait's methods during the `plugins_loaded` hook using `NoteTraits`.
  - Run the unit tests using `npm run test:php`, or
  - Place this code in `mu-plugins`:

  ```php
  add_action( 'plugins_loaded', function () {
  	// Disable WC Admin. If the line below is uncommented, no exception is thrown.
  	add_filter( 'woocommerce_admin_disabled','__return_true' );
  
  	// Try calling a NoteTraits method.
  	( new TestNoteTraitsUser() )->add_note();
  } );
  
  class TestNoteTraitsUser {
  	use Automattic\WooCommerce\Admin\Notes\NoteTraits;
  
  	// Constant required by NoteTraits.
  	const NOTE_NAME = 'test';
  
  	// Method required by NoteTraits.
  	public static function get_note() {
  		return new Automattic\WooCommerce\Admin\Notes\Note();
  	}
  
  	public function add_note() {
  		self::possibly_add_note();
  	}
  }
  ```
- See that a custom exception is thrown rather than a fatal error.

### Considered alternative approaches

Instead of checking if WC Admin is disabled in every `NoteTraits` method that depends on the `admin-note` data store being present, I also considered doing this:

```php
private static function load_data_store() {
	try {
		return new \WC_Data_Store::load( 'admin-note' );
	} catch ( Exception $e ) {
		throw new NotesDisabledException();
	}
}

// And then, in every method replace this:
$data_store = \WC_Data_Store::load( 'admin-note' );

// With this:
$data_store = self::load_data_store();
```

However, if notes become a feature that can be disabled/enabled (see #6568) or more reasons why note creation can fail appear (other than the data store not being registered), I think the approach suggested in this PR will be more flexible.

### CODE REVIEW

**Comment by moon0326** [comment]
Thank you @waclawjacek 👍  for putting this together. It tested well and the code looks good to me. 

There is an ongoing [discussion](https://wp.me/pcShBQ-1i) on disabling `woocommerce_admin_disabled` filter though. I think the change in this PR is safe, but I also think we might need to consider your alternative approach so that we don't have to make another PR in the future. What do you think?

**Comment by waclawjacek** [comment]
@moon0326 Thank you!

I saw that discussion and I thought that it might become possible to disable/enable notes using a feature flag as a result of it. If this happened, the filter in this PR could be replaced with `woocommerce_admin_features` which is something I kept in mind while working on it.

Rereading that discussion, I see this was probably not the right assumption to make and notes might not become a feature flag. I am absolutely OK with implementing the alternative approach instead and will be happy to replace the current one with it.

If notes become something that is always enabled when WC Admin is running and cannot be disabled using a filter, the changes in this PR wouldn't be as important as they are now, at least not in the current sense of handling the "is WC Admin running but disabled?" scenario. Catching the data store loading exception is (IMHO!) more of a nice-to-have rather than part of handling one of the plugin's possible states as it would become something that should never actually happen, at least compared to the situation the PR currently tries to handle.

In short, we currently do expect notes to sometimes be unavailable so we want to handle this. If notes become something that is always available, and for some reason, they are not available at runtime, something went wrong somewhere and an error is in order.

Should I replace the logic in the PR with the `try-catch` statements on data store loading?

**Comment by moon0326** [comment]
> @moon0326 Thank you!
> 
> I saw that discussion and I thought that it might become possible to disable/enable notes using a feature flag as a result of it. If this happened, the filter in this PR could be replaced with `woocommerce_admin_features` which is something I kept in mind while working on it.
> 
> Rereading that discussion, I see this was probably not the right assumption to make and notes might not become a feature flag. I am absolutely OK with implementing the alternative approach instead and will be happy to replace the current one with it.
> 
> If notes become something that is always enabled when WC Admin is running and cannot be disabled using a filter, the changes in this PR wouldn't be as important as they are now, at least not in the current sense of handling the "is WC Admin running but disabled?" scenario. Catching the data store loading exception is (IMHO!) more of a nice-to-have rather than part of handling one of the plugin's possible states as it would become something that should never actually happen, at least compared to the situation the PR currently tries to handle.
> 
> In short, we currently do expect notes to sometimes be unavailable so we want to handle this. If notes become something that is always available, and for some reason, they are not available at runtime, something went wrong somewhere and an error is in order.
> 
> Then, once notes are something that should always be available, the changes in this PR could maybe be retired? I am not sure about this - maybe it'd be nice to keep the custom exception anyway. What do you think?
> 
> Should I replace the logic in the PR with the `try-catch` statements on data store loading?

Yup, I think the custom exception is great to have 👍 

I like both the original (this PR) and the alternative approaches, but I'm leaning towards the alternative approach (try and catch) since it's most likely we're going to deprecate the filter.

**Comment by waclawjacek** [comment]
Okay! I have updated the PR to reflect the alternative approach.

I have also added the check to other classes using the data store, not just `NoteTraits`. I did this by moving the exception handling logic to `Notes::load_data_store()` and then replacing every attempt to load the notes data store with calls to `Notes::load_data_store()`.

While I added tests for the new code, I did not add tests for the changes in existing classes and I did not add a `@throws` tag to the docblocks (yet).

Since the PR is now catching a data store loading error rather than notifying the consuming code that notes are disabled, I have renamed the exception from `NotesDisabledException` to `NotesUnavailableException`.

Please give the PR a look and let me know what you think and if I should update the tests and docblocks!

**Review by moon0326 — APPROVED** [review]
@waclawjacek Thank you for the change 👍  x 1000. They LGTM and tested well. I was able to create notes without a problem. I was also able to raise the exception by disabling the plugin.

**Comment by waclawjacek** [comment]
@moon0326 Thanks so much! 🙌

I did not really test each case where direct references to the store were replaced with calls to `Notes::load_data_store()` as I wasn't sure if this change will get accepted.

Also, I see I left some references to WC Admin being disabled in the code, even though this is not exactly what the code is protecting against now (the current check makes sure if the store is registered, not if WC Admin is disabled), namely in the changelog, the exception code (`woocommerce_admin_notes_disabled`), and the test method names.

Should I test all the places where `Notes::load_data_store()` is referenced (or write automated tests if possible) and adjust the above mentions of notes being disabled?

**Comment by moon0326** [comment]
> @moon0326 Thanks so much! 🙌
> 
> I did not really test each case where direct references to the store were replaced with calls to `Notes::load_data_store()` as I wasn't sure if this change will get accepted.
> 
> Also, I see I left some references to WC Admin being disabled in the code, even though this is not exactly what the code is protecting against now (the current check makes sure if the store is registered, not if WC Admin is disabled), namely in the changelog, the exception code (`woocommerce_admin_notes_disabled`), and the test method names.
> 
> Should I test all the places where `Notes::load_data_store()` is referenced (or write automated tests if possible) and adjust the above mentions of notes being disabled?

I think it's nice to write tests for all of them, but I don't think it's required in my opinion. ```Notes:: load_data_store``` is simple and the method is covered in the [test](https://github.com/woocommerce/woocommerce-admin/pull/6771/files#diff-2be38972e403677995da8a4ae29fdaad316f44edcc21a3c69f6f59d8ac84c329) already.

**Comment by waclawjacek** [comment]
@moon0326 Got it! I will give the PR one more look on Monday, make the adjustments that seem reasonable, and I'll post the update then. Should I merge the PR myself afterward?

**Comment by moon0326** [comment]
> @moon0326 Got it! I will give the PR one more look on Monday, make the adjustments that seem reasonable, and I'll post the update then. Should I merge the PR myself afterward?

Sounds good 👍  Yup! You can merge it yourself.

**Comment by waclawjacek** [comment]
I have updated the changes so that they mention the inability to load the data store rather than WC Admin being disabled.

I have manually tested each call to `Notes::load_data_store()` by setting up a test site in a way that triggers the conditions. Everything worked as expected so I am merging this PR!
